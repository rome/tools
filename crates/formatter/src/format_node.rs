use crate::token::{LineMode, Token, Tokens};
use serde_json::Value;

pub trait FormatValue {
	fn format(&self) -> Token;
}

impl FormatValue for Value {
	fn format(&self) -> Token {
		match self {
			Value::String(string) => Token::string(format!(r#""{}""#, string).as_str()),
			Value::Number(number) => {
				let number = number.as_u64().unwrap();
				Token::from(number)
			}
			Value::Bool(value) => Token::from(value),
			Value::Object(value) => {
				let mut final_tokens: Tokens = vec![];
				final_tokens.push("{".into());
				let mut content = vec![];
				for (key, value) in value {
					let tokens: Tokens = vec![
						"\"".into(),
						key.as_str().into(),
						"\"".into(),
						":".into(),
						Token::Space,
						value.format(),
						",".into(),
						Token::Break,
					];
					content.push(Token::concat(tokens))
				}
				final_tokens.push(Token::indent(Token::from(content)));
				final_tokens.push("}".into());
				Token::from(final_tokens)
			}
			_ => unimplemented!("Implement rest"),
		}
	}
}
