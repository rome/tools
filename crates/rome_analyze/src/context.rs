use crate::{registry::RuleRoot, services::ServiceBag, Queryable, Rule};
use std::ops::Deref;

pub struct RuleContext<'a, R>
where
    R: ?Sized + Rule,
{
    query_result: &'a <<R as Rule>::Query as Queryable>::Output,
    root: &'a RuleRoot<R>,
    services: <<R as Rule>::Query as Queryable>::Services,
}

impl<'a, R> RuleContext<'a, R>
where
    R: Rule,
{
    pub fn new(
        query_result: &'a <<R as Rule>::Query as Queryable>::Output,
        root: &'a RuleRoot<R>,
        services: ServiceBag,
    ) -> Result<Self, <<<R as Rule>::Query as Queryable>::Services as TryFrom<ServiceBag>>::Error>
    {
        Ok(Self {
            query_result,
            root,
            services: services.try_into()?,
        })
    }

    pub fn query(&self) -> &<<R as Rule>::Query as Queryable>::Output {
        self.query_result
    }

    pub fn root(&self) -> RuleRoot<R> {
        self.root.clone()
    }
}

impl<'a, R> Deref for RuleContext<'a, R>
where
    R: Rule,
{
    type Target = <<R as Rule>::Query as Queryable>::Services;

    fn deref(&self) -> &Self::Target {
        &self.services
    }
}
