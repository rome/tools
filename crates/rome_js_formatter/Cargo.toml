[package]
name = "rome_js_formatter"
version = "0.0.0"
edition = "2021"
authors = ["Rome Tools Developers and Contributors"]
repository = "https://github.com/rome/tools"
license = "MIT"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
cfg-if = "1.0.0"
rome_js_syntax = { path = "../rome_js_syntax" }
rome_js_factory = { path = "../rome_js_factory" }
rome_formatter = { path = "../rome_formatter" }
rome_rowan = { path = "../rome_rowan" }
rome_text_size = { path = "../rome_text_size"}
rome_diagnostics_categories = { path = "../rome_diagnostics_categories"}
tracing = { workspace = true }
unicode-width = "0.1.9"
serde = { version = "1.0.136", features = ["derive"], optional = true }
schemars = { version = "0.8.10", optional = true }

[dev-dependencies]
rome_parser = { path = "../rome_parser" }
rome_formatter_test = {path = "../rome_formatter_test"}
rome_console = { path = "../rome_console" }
rome_fs = { path = "../rome_fs" }
rome_service = { path = "../rome_service" }
rome_js_parser = { path = "../rome_js_parser" }
rome_js_factory = { path = "../rome_js_factory" }
serde = { version = "1", features = ["derive"] }
serde_json = "1"
tests_macros = { path = "../tests_macros" }
insta = { version="1.18.2", features = ["glob"] }
rome_diagnostics = { path = "../rome_diagnostics" }
countme = { workspace = true, features = ["enable"] }
similar = "2.1.0"
similar-asserts = "1.2.0"
quickcheck = "1.0.3"
quickcheck_macros = "1.0.0"
iai = "0.1.1"

[[bench]]
name = "iai"
harness = false

[features]
serde = ["dep:serde", "schemars"]
