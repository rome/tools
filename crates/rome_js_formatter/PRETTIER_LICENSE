This crate contains derivative work from the https://github.com/prettier/prettier project. Notably we have either taken heavy inspiration or copied the semantics and heuristics for code formatting JavaScript. This has been done to provide ecosystem compatibility for what is currently a community standard. Below you will find the LICENSE from the Prettier project:

------------------------------------------------------

Copyright Â© James Long and contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
