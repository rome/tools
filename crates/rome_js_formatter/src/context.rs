use rome_formatter::printer::PrinterOptions;
use rome_formatter::{
    CommentKind, CommentStyle, Comments, CstFormatContext, FormatContext, FormatOptions,
    IndentStyle, LineWidth, TransformSourceMap,
};
use rome_js_syntax::suppression::{parse_suppression_comment, SuppressionCategory};
use rome_js_syntax::{JsLanguage, JsSyntaxKind, SourceType};
use rome_rowan::SyntaxTriviaPieceComments;
use std::fmt;
use std::fmt::Debug;
use std::rc::Rc;
use std::str::FromStr;

#[derive(Debug, Clone, Default)]
pub struct JsFormatContext {
    options: JsFormatOptions,

    /// The comments of the nodes and tokens in the program.
    comments: Rc<Comments<JsLanguage>>,

    source_map: Option<TransformSourceMap>,
}

impl JsFormatContext {
    pub fn new(options: JsFormatOptions, comments: Comments<JsLanguage>) -> Self {
        Self {
            options,
            comments: Rc::new(comments),
            ..JsFormatContext::default()
        }
    }

    pub fn with_source_map(mut self, source_map: Option<TransformSourceMap>) -> Self {
        self.source_map = source_map;
        self
    }
}

#[derive(Eq, PartialEq, Debug, Copy, Clone, Hash)]
pub struct TabWidth(u8);

impl From<u8> for TabWidth {
    fn from(value: u8) -> Self {
        TabWidth(value)
    }
}

impl From<TabWidth> for u8 {
    fn from(width: TabWidth) -> Self {
        width.0
    }
}

impl FormatContext for JsFormatContext {
    type Options = JsFormatOptions;

    fn options(&self) -> &Self::Options {
        &self.options
    }

    fn source_map(&self) -> Option<&TransformSourceMap> {
        self.source_map.as_ref()
    }
}

impl CstFormatContext for JsFormatContext {
    type Language = JsLanguage;
    type Style = JsCommentStyle;

    fn comment_style(&self) -> Self::Style {
        JsCommentStyle
    }

    fn comments(&self) -> Rc<Comments<JsLanguage>> {
        self.comments.clone()
    }
}

#[derive(Debug, Clone, Default)]
pub struct JsFormatOptions {
    /// The indent style.
    indent_style: IndentStyle,

    /// What's the max width of a line. Defaults to 80.
    line_width: LineWidth,

    /// The style for quotes. Defaults to double.
    quote_style: QuoteStyle,

    /// When properties in objects are quoted. Defaults to as-needed.
    quote_properties: QuoteProperties,

    /// Information related to the current file
    source_type: SourceType,
}

impl JsFormatOptions {
    pub fn new(source_type: SourceType) -> Self {
        Self {
            source_type,
            ..JsFormatOptions::default()
        }
    }

    pub fn with_indent_style(mut self, indent_style: IndentStyle) -> Self {
        self.indent_style = indent_style;
        self
    }

    pub fn with_line_width(mut self, line_width: LineWidth) -> Self {
        self.line_width = line_width;
        self
    }

    pub fn with_quote_style(mut self, quote_style: QuoteStyle) -> Self {
        self.quote_style = quote_style;
        self
    }

    pub fn with_quote_properties(mut self, quote_properties: QuoteProperties) -> Self {
        self.quote_properties = quote_properties;
        self
    }

    pub fn with_source_type(mut self, source_type: SourceType) -> Self {
        self.source_type = source_type;
        self
    }

    pub fn quote_style(&self) -> QuoteStyle {
        self.quote_style
    }

    pub fn quote_properties(&self) -> QuoteProperties {
        self.quote_properties
    }

    pub fn source_type(&self) -> SourceType {
        self.source_type
    }

    pub fn tab_width(&self) -> TabWidth {
        match self.indent_style {
            IndentStyle::Tab => 2.into(),
            IndentStyle::Space(quantities) => quantities.into(),
        }
    }
}

impl FormatOptions for JsFormatOptions {
    fn indent_style(&self) -> IndentStyle {
        self.indent_style
    }

    fn line_width(&self) -> LineWidth {
        self.line_width
    }

    fn as_print_options(&self) -> PrinterOptions {
        PrinterOptions::default()
            .with_indent(self.indent_style)
            .with_print_width(self.line_width.into())
    }
}

impl fmt::Display for JsFormatOptions {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "Indent style: {}", self.indent_style)?;
        writeln!(f, "Line width: {}", self.line_width.value())?;
        writeln!(f, "Quote style: {}", self.quote_style)?;
        writeln!(f, "Quote properties: {}", self.quote_properties)
    }
}

#[derive(Eq, PartialEq, Copy, Clone, Debug, Default)]
pub struct JsCommentStyle;

impl CommentStyle<JsLanguage> for JsCommentStyle {
    fn is_suppression(&self, text: &str) -> bool {
        parse_suppression_comment(text)
            .flat_map(|suppression| suppression.categories)
            .any(|(category, _)| category == SuppressionCategory::Format)
    }

    fn get_comment_kind(&self, comment: &SyntaxTriviaPieceComments<JsLanguage>) -> CommentKind {
        if comment.text().starts_with("/*") {
            if comment.has_newline() {
                CommentKind::Block
            } else {
                CommentKind::InlineBlock
            }
        } else {
            CommentKind::Line
        }
    }

    fn is_group_start_token(&self, kind: JsSyntaxKind) -> bool {
        matches!(
            kind,
            JsSyntaxKind::L_PAREN
                | JsSyntaxKind::L_BRACK
                | JsSyntaxKind::L_CURLY
                | JsSyntaxKind::DOLLAR_CURLY
        )
    }

    fn is_group_end_token(&self, kind: JsSyntaxKind) -> bool {
        matches!(
            kind,
            JsSyntaxKind::R_BRACK
                | JsSyntaxKind::R_CURLY
                | JsSyntaxKind::R_PAREN
                | JsSyntaxKind::COMMA
                | JsSyntaxKind::SEMICOLON
                | JsSyntaxKind::DOT
                | JsSyntaxKind::EOF
        )
    }
}

#[derive(Debug, Eq, PartialEq, Clone, Copy)]
#[cfg_attr(
    feature = "serde",
    derive(serde::Serialize, serde::Deserialize, schemars::JsonSchema)
)]
#[derive(Default)]
pub enum QuoteStyle {
    #[default]
    Double,
    Single,
}

impl FromStr for QuoteStyle {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "double" | "Double" => Ok(Self::Double),
            "single" | "Single" => Ok(Self::Single),
            // TODO: replace this error with a diagnostic
            _ => Err("Value not supported for QuoteStyle"),
        }
    }
}

impl fmt::Display for QuoteStyle {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            QuoteStyle::Double => write!(f, "Double Quotes"),
            QuoteStyle::Single => write!(f, "Single Quotes"),
        }
    }
}

impl QuoteStyle {
    pub fn as_char(&self) -> char {
        match self {
            QuoteStyle::Double => '"',
            QuoteStyle::Single => '\'',
        }
    }

    pub fn as_string(&self) -> &str {
        match self {
            QuoteStyle::Double => "\"",
            QuoteStyle::Single => "'",
        }
    }

    /// Returns the quote, prepended with a backslash (escaped)
    pub fn as_escaped(&self) -> &str {
        match self {
            QuoteStyle::Double => "\\\"",
            QuoteStyle::Single => "\\'",
        }
    }

    pub fn as_bytes(&self) -> u8 {
        self.as_char() as u8
    }

    /// Returns the quote in HTML entity
    pub fn as_html_entity(&self) -> &str {
        match self {
            QuoteStyle::Double => "&quot;",
            QuoteStyle::Single => "&apos;",
        }
    }

    /// Given the current quote, it returns the other one
    pub fn other(&self) -> Self {
        match self {
            QuoteStyle::Double => QuoteStyle::Single,
            QuoteStyle::Single => QuoteStyle::Double,
        }
    }
}

#[derive(Debug, Eq, PartialEq, Clone, Copy)]
#[cfg_attr(
    feature = "serde",
    derive(serde::Serialize, serde::Deserialize, schemars::JsonSchema)
)]
pub enum QuoteProperties {
    AsNeeded,
    Preserve,
}

impl Default for QuoteProperties {
    fn default() -> Self {
        Self::AsNeeded
    }
}

impl FromStr for QuoteProperties {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "as-needed" | "AsNeeded" => Ok(Self::AsNeeded),
            "preserve" | "Preserve" => Ok(Self::Preserve),
            // TODO: replace this error with a diagnostic
            _ => Err("Value not supported for QuoteProperties"),
        }
    }
}

impl fmt::Display for QuoteProperties {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            QuoteProperties::AsNeeded => write!(f, "As needed"),
            QuoteProperties::Preserve => write!(f, "Preserve"),
        }
    }
}
