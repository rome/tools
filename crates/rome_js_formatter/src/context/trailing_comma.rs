use crate::prelude::*;
use crate::{JsFormatContext, JsFormatOptions};
use rome_deserialize::json::with_only_known_variants;
use rome_deserialize::{DeserializationDiagnostic, VisitNode};
use rome_formatter::formatter::Formatter;
use rome_formatter::prelude::{if_group_breaks, text};
use rome_formatter::write;
use rome_formatter::{Format, FormatResult};
use rome_json_syntax::JsonLanguage;
use rome_rowan::SyntaxNode;
use std::fmt;
use std::str::FromStr;

/// This enum is used within formatting functions to print or omit trailing comma.
#[derive(Debug, Copy, Clone)]
pub(crate) enum FormatTrailingComma {
    /// Print trailing comma if the option is [TrailingComma::All].
    All,
    /// Print trailing comma if the option is [TrailingComma::All] or [TrailingComma::Es5].
    ES5,
}

impl FormatTrailingComma {
    /// This function returns corresponding [TrailingSeparator] for [format_separated] function.
    pub fn trailing_separator(&self, options: &JsFormatOptions) -> TrailingSeparator {
        if options.trailing_comma.is_none() {
            return TrailingSeparator::Omit;
        }

        match self {
            FormatTrailingComma::All => {
                if options.trailing_comma.is_all() {
                    TrailingSeparator::Allowed
                } else {
                    TrailingSeparator::Omit
                }
            }
            FormatTrailingComma::ES5 => TrailingSeparator::Allowed,
        }
    }
}

impl Format<JsFormatContext> for FormatTrailingComma {
    fn fmt(&self, f: &mut Formatter<JsFormatContext>) -> FormatResult<()> {
        if f.options().trailing_comma.is_none() {
            return Ok(());
        }

        if matches!(self, FormatTrailingComma::ES5) || f.options().trailing_comma().is_all() {
            write!(f, [if_group_breaks(&text(","))])?
        }

        Ok(())
    }
}

/// Print trailing commas wherever possible in multi-line comma-separated syntactic structures.
#[derive(Default, Debug, Eq, PartialEq, Clone, Copy)]
#[cfg_attr(
    feature = "serde",
    derive(serde::Serialize, serde::Deserialize, schemars::JsonSchema),
    serde(rename_all = "camelCase")
)]
pub enum TrailingComma {
    /// Trailing commas wherever possible (including function parameters and calls).
    #[default]
    All,
    /// Trailing commas where valid in ES5 (objects, arrays, etc.). No trailing commas in type parameters in TypeScript.
    Es5,
    /// No trailing commas.
    None,
}

impl TrailingComma {
    pub(crate) const KNOWN_VALUES: &'static [&'static str] = &["all", "es5", "none"];

    pub const fn is_es5(&self) -> bool {
        matches!(self, TrailingComma::Es5)
    }
    pub const fn is_all(&self) -> bool {
        matches!(self, TrailingComma::All)
    }
    pub const fn is_none(&self) -> bool {
        matches!(self, TrailingComma::None)
    }
}

impl FromStr for TrailingComma {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "es5" | "ES5" => Ok(Self::Es5),
            "all" | "All" => Ok(Self::All),
            "none" | "None" => Ok(Self::None),
            // TODO: replace this error with a diagnostic
            _ => Err("Value not supported for TrailingComma"),
        }
    }
}

impl fmt::Display for TrailingComma {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            TrailingComma::Es5 => std::write!(f, "ES5"),
            TrailingComma::All => std::write!(f, "All"),
            TrailingComma::None => std::write!(f, "None"),
        }
    }
}

impl VisitNode<JsonLanguage> for TrailingComma {
    fn visit_member_value(
        &mut self,
        node: &SyntaxNode<JsonLanguage>,
        diagnostics: &mut Vec<DeserializationDiagnostic>,
    ) -> Option<()> {
        let node = with_only_known_variants(node, TrailingComma::KNOWN_VALUES, diagnostics)?;
        match node.inner_string_text().ok()?.text() {
            "all" => {
                *self = TrailingComma::All;
            }
            "es5" => {
                *self = TrailingComma::Es5;
            }
            "none" => {
                *self = TrailingComma::None;
            }
            _ => {}
        }
        Some(())
    }
}
