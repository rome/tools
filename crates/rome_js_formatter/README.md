# `rome_js_formatter`

The official formatter used by Rome.

## Internal installation

```toml
rome_js_formatter = { version = "0.0.0", path = "../rome_js_formatter" }
```

## Usage

The foundation of the formatter relies on two pillars:

- the usage of the [*trait*](https://doc.rust-lang.org/reference/items/traits.html) generic `Format` trait and `FormatNode` for nodes.
- the creation of an intermediate IR via a series of helpers

Import the `FormatNode` trait and implement it for your Node.

```rust
use rome_js_formatter::{Format, FormatNode, FormatElement, format_elements, token};

struct Buzz {
 blast: String
}

impl FormatNode for Buzz {
 fn format_fields(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
 	// implementation goes here
 	format_elements![token("_"), blast.as_str(), token("_")]
 }
}

```

## Rules when formatting AST nodes

1. if a token is mandatory and the AST has that information, please use that token instead, for example:

	```rust
	fn format_fields(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
		let l_paren_yes = &self.l_paren_token()?.format(); // yes
		let l_paren_no = toke("("); // no
	}
	```

 1. for tokens that are not mandatory, use our helpers
 1. do not attempt to "fix" the code. If you know a token/node is mandatory, return `None` instead
