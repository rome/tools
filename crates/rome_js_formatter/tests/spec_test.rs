use rome_diagnostics::file::FileId;
use rome_formatter::{FormatOptions, LineWidth};
use rome_formatter::{IndentStyle, Printed};
use rome_fs::RomePath;
use rome_js_formatter::context::{JsFormatOptions, QuoteProperties, QuoteStyle};
use rome_js_formatter::format_node;
use rome_js_parser::parse;
use rome_js_syntax::{ModuleKind, SourceType};
use rome_service::workspace::{FeatureName, SupportsFeatureParams};
use rome_service::App;
use serde::{Deserialize, Serialize};
use std::fmt::Debug;
use std::fmt::Write;
use std::fs;
use std::path::{Path, PathBuf};

mod check_reformat {
    include!("check_reformat.rs");
}

#[derive(Debug, Eq, PartialEq, Clone, Copy, Deserialize, Serialize)]
pub enum SerializableIndentStyle {
    /// Tab
    Tab,
    /// Space, with its quantity
    Space(u8),
}

impl From<SerializableIndentStyle> for IndentStyle {
    fn from(test: SerializableIndentStyle) -> Self {
        match test {
            SerializableIndentStyle::Tab => IndentStyle::Tab,
            SerializableIndentStyle::Space(spaces) => IndentStyle::Space(spaces),
        }
    }
}

#[derive(Debug, Eq, PartialEq, Clone, Copy, Deserialize, Serialize)]
pub enum SerializableQuoteStyle {
    Double,
    Single,
}

impl From<SerializableQuoteStyle> for QuoteStyle {
    fn from(test: SerializableQuoteStyle) -> Self {
        match test {
            SerializableQuoteStyle::Double => QuoteStyle::Double,
            SerializableQuoteStyle::Single => QuoteStyle::Single,
        }
    }
}

#[derive(Debug, Eq, PartialEq, Clone, Copy, Deserialize, Serialize)]
pub enum SerializableQuoteProperties {
    AsNeeded,
    Preserve,
}

impl From<SerializableQuoteProperties> for QuoteProperties {
    fn from(test: SerializableQuoteProperties) -> Self {
        match test {
            SerializableQuoteProperties::AsNeeded => QuoteProperties::AsNeeded,
            SerializableQuoteProperties::Preserve => QuoteProperties::Preserve,
        }
    }
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy)]
pub struct SerializableFormatOptions {
    /// The indent style.
    pub indent_style: Option<SerializableIndentStyle>,

    /// What's the max width of a line. Defaults to 80.
    pub line_width: Option<u16>,

    // The style for quotes. Defaults to double.
    pub quote_style: Option<SerializableQuoteStyle>,

    /// When properties in objects are quoted. Defaults to as-needed.
    pub quote_properties: Option<SerializableQuoteProperties>,
}

impl From<SerializableFormatOptions> for JsFormatOptions {
    fn from(test: SerializableFormatOptions) -> Self {
        Self::new(SourceType::default())
            .with_indent_style(
                test.indent_style
                    .map_or_else(|| IndentStyle::Tab, |value| value.into()),
            )
            .with_line_width(
                test.line_width
                    .and_then(|width| LineWidth::try_from(width).ok())
                    .unwrap_or_default(),
            )
            .with_quote_style(
                test.quote_style
                    .map_or_else(|| QuoteStyle::Double, |value| value.into()),
            )
            .with_quote_properties(
                test.quote_properties
                    .map_or_else(|| QuoteProperties::AsNeeded, |value| value.into()),
            )
    }
}

#[derive(Debug, Deserialize, Serialize)]
struct TestOptions {
    cases: Vec<SerializableFormatOptions>,
}

#[derive(Debug, Default)]
struct SnapshotContent {
    input: String,
    output: Vec<(String, JsFormatOptions)>,
}

impl SnapshotContent {
    fn add_output(&mut self, formatted: Printed, options: JsFormatOptions) {
        let code = formatted.as_code();
        let mut output: String = code.to_string();
        if !formatted.verbatim_ranges().is_empty() {
            output.push_str("\n\n");
            output.push_str("## Unimplemented nodes/tokens");
            output.push_str("\n\n");
            for (range, text) in formatted.verbatim() {
                writeln!(output, "{:?} => {:?}", text, range).unwrap();
            }
        }

        let line_width_limit = options.line_width().value() as usize;
        let mut exceeding_lines = code
            .lines()
            .enumerate()
            .filter(|(_, line)| line.len() > line_width_limit)
            .peekable();

        if exceeding_lines.peek().is_some() {
            write!(
                output,
                "\n\n## Lines exceeding width of {line_width_limit} characters\n\n"
            )
            .unwrap();

            for (line_index, text) in exceeding_lines {
                let line_number = line_index + 1;
                writeln!(output, "{line_number:>5}: {text}").unwrap();
            }
        }

        self.output.push((output, options));
    }

    fn set_input(&mut self, content: impl Into<String>) {
        self.input = content.into();
    }

    fn snap_content(&mut self) -> String {
        let mut output = String::new();
        output.push_str("# Input");
        output.push('\n');
        output.push_str(self.input.as_str());
        output.push_str("\n=============================\n");

        output.push_str("# Outputs\n");
        let iter = self.output.iter();
        for (index, (content, options)) in iter.enumerate() {
            let formal_index = index + 1;
            writeln!(output, "## Output {formal_index}").unwrap();
            output.push_str("-----\n");
            write!(output, "{}", options).unwrap();
            output.push_str("-----\n");
            output.push_str(content.as_str());
        }

        output
    }
}

/// [insta.rs](https://insta.rs/docs) snapshot testing
///
/// For better development workflow, run
/// `cargo watch -i '*.new' -x 'test -p rome_js_formatter formatter'`
///
/// To review and commit the snapshots, `cargo install cargo-insta`, and run
/// `cargo insta review` or `cargo insta accept`
///
/// The input and the expected output are stored as dedicated files in the `tests/specs` directory where
/// the input file name is `{spec_name}.json` and the output file name is `{spec_name}.json.snap`.
///
/// Specs can be grouped in directories by specifying the directory name in the spec name. Examples:
///
/// # Examples
///
/// * `json/null` -> input: `tests/specs/json/null.json`, expected output: `tests/specs/json/null.json.snap`
/// * `null` -> input: `tests/specs/null.json`, expected output: `tests/specs/null.json.snap`
pub fn run(spec_input_file: &str, _expected_file: &str, test_directory: &str, file_type: &str) {
    let app = App::default();

    let file_path = &spec_input_file;
    let spec_input_file = Path::new(spec_input_file);

    assert!(
        spec_input_file.is_file(),
        "The input '{}' must exist and be a file.",
        spec_input_file.display()
    );

    let mut rome_path = RomePath::new(file_path, FileId::zero());
    let can_format = app
        .workspace
        .supports_feature(SupportsFeatureParams {
            path: rome_path.clone(),
            feature: FeatureName::Format,
        })
        .unwrap();

    if can_format.reason.is_none() {
        let mut snapshot_content = SnapshotContent::default();
        let buffer = rome_path.get_buffer_from_file();
        let mut source_type: SourceType = rome_path.as_path().try_into().unwrap();
        if file_type != "module" {
            source_type = source_type.with_module_kind(ModuleKind::Script);
        }

        let input = fs::read_to_string(file_path).unwrap();
        snapshot_content.set_input(input.as_str());

        let parsed = parse(buffer.as_str(), FileId::zero(), source_type);
        let has_errors = parsed.has_errors();
        let root = parsed.syntax();

        // we ignore the error for now
        let options = JsFormatOptions::new(source_type);
        let formatted = format_node(options.clone(), &root).unwrap();
        let printed = formatted.print().unwrap();
        let file_name = spec_input_file.file_name().unwrap().to_str().unwrap();

        if !has_errors {
            check_reformat::check_reformat(check_reformat::CheckReformatParams {
                root: &root,
                text: printed.as_code(),
                source_type,
                file_name,
                options: options.clone(),
            });
        }

        snapshot_content.add_output(printed, options);

        let test_directory = PathBuf::from(test_directory);
        let options_path = test_directory.join("options.json");
        if options_path.exists() {
            {
                let mut options_path = RomePath::new(&options_path, FileId::zero());
                // SAFETY: we checked its existence already, we assume we have rights to read it
                let options: TestOptions =
                    serde_json::from_str(options_path.get_buffer_from_file().as_str()).unwrap();

                for test_case in options.cases {
                    let format_options: JsFormatOptions = test_case.into();
                    // we don't track the source type inside the serializable structs, so we
                    // inject it here
                    let formatted = format_node(format_options.clone(), &root).unwrap();
                    let printed = formatted.print().unwrap();

                    if !has_errors {
                        check_reformat::check_reformat(check_reformat::CheckReformatParams {
                            root: &root,
                            text: printed.as_code(),
                            source_type,
                            file_name,
                            options: format_options.clone(),
                        });
                    }

                    snapshot_content.add_output(printed, format_options);
                }
            }
        }

        insta::with_settings!({
            prepend_module_to_snapshot => false,
            snapshot_path => spec_input_file.parent().unwrap(),
        }, {
            insta::assert_snapshot!(file_name, snapshot_content.snap_content(), file_name);
        });
    }
}
