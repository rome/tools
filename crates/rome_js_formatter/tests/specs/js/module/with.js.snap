---
source: crates/rome_js_formatter/tests/spec_test.rs
expression: with.js
---
# Input
with (   b)

{
	5
}

=============================
# Outputs
## Output 1
-----
Indent style: Tab
Line width: 80
Quote style: Double Quotes
Preserve quotes: false
-----
with (   b)

{
	5
}

